overlord/ifacestate/handlers.go

```diff
@@ -327,30 +342,43 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
-	err = m.repo.Disconnect(plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name)
-	if err != nil {
-		return err
-	}
-
-	plug := m.repo.Plug(plugRef.Snap, plugRef.Name)
-	var plugSnapst snapstate.SnapState
-	if err := snapstate.Get(st, plugRef.Snap, &plugSnapst); err != nil {
-		return err
-	}
-	slot := m.repo.Slot(slotRef.Snap, slotRef.Name)
-	var slotSnapst snapstate.SnapState
-	if err := snapstate.Get(st, slotRef.Snap, &slotSnapst); err != nil {
-		return err
+	var affectedConns []interfaces.ConnRef
+	if plugRef.Snap != "" && plugRef.Name != "" && slotRef.Snap != "" && slotRef.Name != "" {
+		if err := m.repo.Disconnect(plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name); err != nil {
+			return err
+		}
+		affectedConns = []interfaces.ConnRef{{*plugRef, *slotRef}}
+	} else if plugRef.Name != "" {
```

Comment:
This would result in an error from repo but I can easily make the test more obvious. I think you are also right since we are multiplexing to one of two functions depending on argument layout.

Done, have a look please.